source/base/Promise.ooc

```diff
@@ -11,53 +12,59 @@ Promise: abstract class {
 	_state : _PromiseState
 	init: func
 	wait: abstract func -> Bool
+	wait: abstract func ~timeout (seconds: Double) -> Bool
 	cancel: virtual func -> Bool { false }
 	start: static func (action: Func) -> This {
 		_ThreadPromise new(action)
 	}
+	operator + (other: This) -> PromiseCollector {
+		collector := PromiseCollector new()
+		collector add(this)
+		collector add(other)
+		collector
+	}
 }
 
 _ThreadPromise: class extends Promise {
 	_action: Func
 	_thread: Thread
-	_mutex := Mutex new()
+	_threadAlive := true
 	init: func (task: Func) {
 		super()
 		this _state = _PromiseState Unfinished
 		this _action = func {
 			task()
-			this _mutex lock()
 			if (this _state != _PromiseState Cancelled)
 				this _state = _PromiseState Finished
-			this _mutex unlock()
 		}
 		this _thread = Thread new(this _action)
 		this _thread start()
 	}
 	free: override func {
-		if (this _state == _PromiseState Unfinished)
+		if (this _threadAlive)
 			this _thread wait()
```

Comment:
The thread doesn't have to be joined, but if you don't join it before the process exits then Valgrind will see it as a memory leak. You could simply free the object data for ThreadPromise and Thread which can be done from inside the thread itself, but I'm not sure if this is compatible with every other platform.

Just leave it for now then, maybe add a TODO?